// Screen/Printer/Enquirer.mjs
import registry from 'registry';

import chalk  from 'chalk';
import enquirer from 'enquirer';

export default class Enquirer {

  #session = null;

  constructor(session){
    this.#session = session;
  }

  async prompt(){
    const avatar = this.#session.user.avatar;

    const prompt = new enquirer.AutoComplete({
      name: 'action',
      message: 'Action>',
      limit: 10,
      choices: avatar.getActionCompletions()
    });
    const input = await prompt.run();

    await avatar.processInput(input);

    // check is there are any other prompts////


    await this.prompt();
  }


  async print(data){
    const user = this.#session.user;
    const avatar = this.#session.user.avatar;

    if(data.type == 'alert'){
      if(data.kind == 'info'){
        console.log(`${chalk.white.bgBlue(user.username+' INFO')}${chalk.blue('\uE0B0')} ${data.text}`);
      }else if(data.kind == 'warning'){
        console.log(`${chalk.red.bgYellow(user.username+' WARNING')}${chalk.yellow('\uE0B0')} ${data.text}`);
      }else if(data.kind == 'error'){
        console.log(`${chalk.black.bgRed(user.username+' ERROR')}${chalk.red('\uE0B0')} ${data.text}`);
      }else{
        console.log(data.text);
      }


    }else{
      console.log(data);
    }

  }

  async input({message}){
    const prompt = new enquirer.Input({
      message: message,
      initial: ''
    });
    const value = await prompt.run();
    return {value};
  }

  async select(data){
    const readline = this.readline;
    const {message, choices} = data;

    const prompt = new enquirer.Select({
      name: 'value',
      message: message,
      choices,
    });

    const value = await prompt.run();
    return {value};
  }

}
