// Screen/Printer/Json.mjs
import registry from 'registry';

export default class Json {

  #session = null;

  constructor(session){
    this.#session = session;
  }

  async prompt(){
    const avatar = this.#session.user.avatar;
    const username = this.#session.username;
    const response = {
      type: 'input',
      name: 'text',
      message: username,
    };
    this.#session.socket.emit('screen', response)
    return response;
  }

  async print(data){
    const username = this.#session.username;
    const avatar = this.#session.user.avatar;
    const response = Object.assign({username},data);
    this.#session.socket.emit('screen', response)
    return response;
  }

  async input({message}){
    const response = {
      message: message,
      initial: ''
    };
    this.#session.socket.emit('screen', response)
    return response;
  }

  async select({message, choices}){
    const response = {
      name: 'value',
      message: message,
      choices,
    };
    this.#session.socket.emit('screen', response)
    return response;
  }

}
